src/http/gateway/resolver.js

```diff
@@ -0,0 +1,104 @@
+'use strict'
+
+const mh = require('multihashes')
+const promisify = require('promisify-es6')
+const eachOfSeries = require('async/eachOfSeries')
+const debug = require('debug')
+const log = debug('jsipfs:http-gateway:resolver')
+log.error = debug('jsipfs:http-gateway:resolver:error')
+
+const html = require('./utils/html')
+const PathUtil = require('./utils/path')
+
+const INDEX_HTML_FILES = [ 'index.html', 'index.htm', 'index.shtml' ]
+
+function noop () {}
+
+const resolveDirectory = promisify((ipfs, path, multihash, callback) => {
+  callback = callback || noop
+
+  mh.validate(mh.fromB58String(multihash))
+
+  ipfs.object.get(multihash, { enc: 'base58' }, (err, dagNode) => {
+    if (err) { return callback(err) }
+
+    const links = dagNode.links
+    const indexFiles = links.filter((link) => INDEX_HTML_FILES.indexOf(link.name) !== -1)
+
+    // found index file in links
+    if (indexFiles.length > 0) {
+      return callback(null, indexFiles)
+    }
+
+    return callback(null, html.build(path, links))
+  })
+})
+
+const resolveMultihash = promisify((ipfs, path, callback) => {
+  callback = callback || noop
+
+  const parts = PathUtil.splitPath(path)
+  const partsLength = parts.length
+
+  let currentMultihash = parts[0]
+
+  eachOfSeries(parts, (multihash, currentIndex, next) => {
+    // throws error when invalid multihash is passed
+    mh.validate(mh.fromB58String(currentMultihash))
+    log('currentMultihash: ', currentMultihash)
+    log('currentIndex: ', currentIndex, '/', partsLength)
+
+    ipfs.object.get(currentMultihash, { enc: 'base58' }, (err, dagNode) => {
+      if (err) { return next(err) }
+
+      if (currentIndex === partsLength - 1) {
+        // leaf node
+        log('leaf node: ', currentMultihash)
+
+        // TODO: Check if it is a directory by using Unixfs Type, right now
+        // it won't detect empty dirs
```

Comment:
Should be checking to see if it's https://github.com/ipfs/js-ipfs-unixfs/blob/master/src/unixfs.proto.js#L6

Otherwise, it will fail for empty dirs or misbehave for shards.